sql/mssql/diff.go

```diff
@@ -17,51 +22,300 @@ var DefaultDiff schema.Differ = &sqlx.Diff{DiffDriver: &diff{}}
+		return schema.NoChange, err
+	}
+	if changed {
+		change |= schema.ChangeDefault
+	}
+	if identityChanged(from.Attrs, to.Attrs) {
+		change |= schema.ChangeAttr
+	}
+	if changed, err = generatedChanged(from, to); err != nil {
+		return schema.NoChange, err
+	}
+	if changed {
+		change |= schema.ChangeGenerated
+	}
+	return change, nil
 }
 
 // IsGeneratedIndexName reports if the index name was generated by the database.
-func (d *diff) IsGeneratedIndexName(_ *schema.Table, _ *schema.Index) bool {
-	// Not implemented yet.
-	return false
+func (d *diff) IsGeneratedIndexName(t *schema.Table, i *schema.Index) bool {
+	// MS-SQL generates index names for primary keys.
+	return fmt.Sprintf("PK_%s", t.Name) == i.Name
 }
 
 // IndexAttrChanged reports if the index attributes were changed.
-// The default type is BTREE if no type was specified.
-func (*diff) IndexAttrChanged(_, _ []schema.Attr) bool {
-	// Not implemented yet.
+// The default type is CLUSTERED if no type was specified.
+func (*diff) IndexAttrChanged(from, to []schema.Attr) bool {
+	t1, t2 := indexType(from), indexType(to)
+	if t1.T != t2.T {
+		return true
+	}
+	var p1, p2 IndexPredicate
+	if sqlx.Has(from, &p1) != sqlx.Has(to, &p2) || (p1.P != p2.P && p1.P != sqlx.MayWrap(p2.P)) {
+		return true
+	}
+	if indexIncludeChanged(from, to) {
+		return true
+	}
 	return false
 }
 
 // IndexPartAttrChanged reports if the index-part attributes were changed.
 func (*diff) IndexPartAttrChanged(_, _ *schema.Index, _ int) bool {
-	// Not implemented yet.
 	return false
 }
 
 // ReferenceChanged reports if the foreign key referential action was changed.
-func (*diff) ReferenceChanged(_, _ schema.ReferenceOption) bool {
-	// Not implemented yet.
+func (*diff) ReferenceChanged(from, to schema.ReferenceOption) bool {
+	// According to MS-SQL, the NO ACTION rule is set
+	// if no referential action was defined in foreign key.
+	if from == "" {
+		from = schema.NoAction
+	}
+	if to == "" {
+		to = schema.NoAction
+	}
+	return from != to
+}
+
+// collationChange returns the schema change for migrating the collation if
+// it was changed, and it is not the default attribute inherited from its parent.
+func collationChange(from, top, to []schema.Attr) schema.Change {
+	var fromC, topC, toC schema.Collation
+	switch fromHas, topHas, toHas := sqlx.Has(from, &fromC), sqlx.Has(top, &topC), sqlx.Has(to, &toC); {
+	case !fromHas && !toHas:
+	case !fromHas:
+		return &schema.AddAttr{
+			A: &toC,
+		}
+	case !toHas:
+		// There is no way to DROP a COLLATE that was configured on the table,
+		// and it is not the default. Therefore, we use ModifyAttr and give it
+		// the inherited (and default) collation from schema or server.
+		if topHas && fromC.V != topC.V {
+			return &schema.ModifyAttr{
+				From: &fromC,
+				To:   &topC,
+			}
+		}
+	case fromC.V != toC.V:
+		return &schema.ModifyAttr{
+			From: &fromC,
+			To:   &toC,
+		}
+	}
+	return noChange
+}
```

Comment:
Did you verify this behavior in SQL Server? Because the code was taken from the current drivers.

Reply:
Checked with their document, SQL Server has three objects that has collation attribute are: server, database, and column. And because we only managed schema, table and column. Only thing we can update here is the column.